third_party/move/move-vm/runtime/src/data_cache.rs

```diff
@@ -128,108 +199,241 @@ impl TransactionDataCache {
-                GlobalValue::cached(val)?
+        let (entry, bytes_loaded) = match existing_entry {
+            Entry::Vacant(vacant_entry) => {
+                // Nothing is cached: charge for size and potentially bytes
+                let struct_tag = match module_storage.runtime_environment().ty_to_ty_tag(ty)? {
+                    TypeTag::Struct(struct_tag) => *struct_tag,
+                    _ => {
+                        // Since every resource is a struct, the tag must be also a struct tag.
+                        return Err(PartialVMError::new(StatusCode::INTERNAL_TYPE_ERROR));
+                    },
+                };
+
+                let (layout, contains_delayed_fields) = layout_converter
+                    .type_to_type_layout_with_delayed_fields(&mut gas_meter, traversal_context, ty)?
+                    .unpack();
+
+                let (cached_info, num_bytes_loaded) = create_and_charge_entry(
+                    &mut gas_meter,
+                    traversal_context,
+                    &struct_tag,
+                    &layout,
+                    contains_delayed_fields,
+                    false,
+                )?;
+
+                let new_entry = DataCacheEntry {
+                    struct_tag,
+                    layout,
+                    contains_delayed_fields,
+                    value: cached_info,
+                };
+
+                (vacant_entry.insert(new_entry), num_bytes_loaded)
+            },
+            Entry::Occupied(mut occupied_entry) => {
+                // If entry already exists we might only need to upgrade it from SizeOnly to Value and charge for bytes
+                if load_data && !matches!(occupied_entry.get().value, CachedInformation::Value(_)) {
+                    let v = occupied_entry.get();
+                    let (cached_info, _) = create_and_charge_entry(
+                        &mut gas_meter,
+                        traversal_context,
+                        &v.struct_tag,
+                        &v.layout,
+                        v.contains_delayed_fields,
+                        true,
+                    )?;
+                    occupied_entry.get_mut().value = cached_info;
+                }
+                (occupied_entry.into_mut(), NumBytes::zero())
             },
-            None => GlobalValue::none(),
         };
 
-        let entry = DataCacheEntry {
-            struct_tag,
-            layout,
-            contains_delayed_fields,
-            value,
-        };
-        Ok((entry, NumBytes::new(bytes_loaded as u64)))
+        Ok((&entry.value, bytes_loaded))
     }
 
-    /// Returns true if resource has been inserted into the cache. Otherwise, returns false. The
-    /// state of the cache does not chang when calling this function.
-    pub(crate) fn contains_resource(&self, addr: &AccountAddress, ty: &Type) -> bool {
+    fn find_entry(&self, addr: &AccountAddress, ty: &Type) -> Option<&DataCacheEntry> {
```

Comment:
nit: if we change this to call entry, we may be able to also re-use when getting existing_entry in create_and_insert_or_upgrade_and_charge_data_cache_entry.